numpy/core/src/_simd/_simd_inc.h.src

```diff
@@ -0,0 +1,404 @@
+    /**end repeat**/
+    // vectors
+    /**begin repeat
+     * #sfx = u8, u16, u32, u64, s8, s16, s32, s64, f32, f64, b8, b16, b32, b64#
+     */
+    simd_data_v@sfx@,
+    /**end repeat**/
+    // multi-vectors x2
+    /**begin repeat
+     * #sfx = u8, u16, u32, u64, s8, s16, s32, s64, f32, f64#
+     */
+    simd_data_v@sfx@x2,
+    /**end repeat**/
+    // multi-vectors x3
+    /**begin repeat
+     * #sfx = u8, u16, u32, u64, s8, s16, s32, s64, f32, f64#
+     */
+    simd_data_v@sfx@x3,
+    /**end repeat**/
+    simd_data_end,
+} simd_data_type;
+/************************************
+ ** Declarations (inc_data)
+ ************************************/
+/**
+ * simd_data_type information
+ */
+typedef struct
+{
+    // type name compatible with python style
+    const char *pyname;
+    // returns '1' if the type represent a unsigned integer
+    int is_unsigned:1;
+    // returns '1' if the type represent a signed integer
+    int is_signed:1;
+    // returns '1' if the type represent a single or double precision
+    int is_float:1;
+    // returns '1' if the type represent a boolean
+    int is_bool:1;
+    // returns '1' if the type represent a sequence
+    int is_sequence:1;
+    // returns '1' if the type represent a scalar
+    int is_scalar:1;
+    // returns '1' if the type represent a vector
+    int is_vector:1;
+    // returns the len of multi-vector if the type reprsent x2 or x3 vector
+    // otherwise returns 0, e.g. returns 2 if data type is simd_data_vu8x2
+    int is_vectorx;
+    // returns the equivalent scalar data type e.g. simd_data_vu8 -> simd_data_u8
+    simd_data_type to_scalar;
+    // returns the equivalent scalar data type e.g. simd_data_s8 -> simd_data_vs8
+    // NOTE: returns the will equivalent "unsigned" vector type in case of "boolean" vector
+    // e.g. simd_data_vb8 -> simd_data_vu8
+    simd_data_type to_vector;
+    // number of vector lanes
+    int nlanes;
+    // sizeof lane type
+    int lane_size;
+} simd_data_info;
+
+/**
+ * Returns data info of certain dtype.
+ *
+ * Example:
+ **  const simd_data_info *info = simd_data_getinfo(simd_data_vu8);
+ **  if (info->is_vector && info->is_unsigned) {
+ **     ...
+ **  }
+ */
+static const simd_data_info *
+simd_data_getinfo(simd_data_type dtype);
+
+/************************************
+ ** Declarations (inc_vector)
+ ************************************/
+typedef struct
+{
+    PyObject_HEAD
+    // vector type id
+    simd_data_type type;
+    // vector data, aligned for safe casting
+    npyv_lanetype_u8 NPY_DECL_ALIGNED(NPY_SIMD_WIDTH) data[NPY_SIMD_WIDTH];
+} simd_vector;
+/**
+ * convert simd_data to PyObject(simd_vector),
+ * raise Python exception on failure and returns NULL.
+ */
+static simd_vector *
+simd_vector_to_obj(simd_data data, simd_data_type vtype);
+/**
+ * convert PyObject(simd_vector) to simd_data,
+ * raise Python exception on failure.
+ */
+static simd_data
+simd_vector_from_obj(simd_vector *vec, simd_data_type vtype);
```

Comment:
These function names are jumbled, and should be either
```suggestion
/**
 * convert simd_data to PyObject(simd_vector),
 * raise Python exception on failure and returns NULL.
 */
static simd_vector *
simd_vector_from_data(simd_data data, simd_data_type vtype);
/**
 * convert PyObject(simd_vector) to simd_data,
 * raise Python exception on failure.
 */
static simd_data
simd_vector_to_data(simd_vector *vec, simd_data_type vtype);
```
or
```suggestion
/**
 * convert simd_data to PyObject(simd_vector),
 * raise Python exception on failure and returns NULL.
 */
static simd_vector *
simd_data_to_obj(simd_data data, simd_data_type vtype);
/**
 * convert PyObject(simd_vector) to simd_data,
 * raise Python exception on failure.
 */
static simd_data
simd_data_from_obj(simd_vector *vec, simd_data_type vtype);
```

Reply:
> These function names are jumbled, and should be either


those conversion functions work through python type `simd_vector`, I will improve comments instead.

Reply:
I don't understand your comment. `pyobj = simd_vector_to_obj(data, dtype)` does not follow the precedent set by  `pyobj = PyInt_FromLong(data)`. `pyobj = simd_vector_from_data(data, dtype)` _would_ match that precedent.

Reply:
Those two functions is used to convert between NPYV vector data types e.g. `npyv_f32` and ~`simd_vector`~ `simd_vector_obj`.
I improved the comments, it should be more understandable.

Reply:
Right - the convention in python is that if converting between a C type `CType` and a python type `PyType`, the functions should be called `PyType_FromCType` and `PyType_ToCType`. What you've done here is `PyType_to_obj` and `PyType_from_obj`, which completely drops the information about which c type is being used.

Reply:
Python API uses a suffix for each C data type, we have dozens of data types here.

> which completely drops CType from the name.

yes but still readable through data type id. e.g. `simd_vector_to_obj(data, simd_data_vu8)`,
plus this approach is more friendly with `C macros` in `_simd_inc_easyintrin.h`.